### ISSUE TITLE
postcss-import does not work with postcss 4.0

### ISSUE THREAD

**Issue by nullstyle** [postcss/postcss-import#16]
I get the following error:

```
/Users/nullstyle/src/tmp/busted/node_modules/postcss-import/node_modules/postcss-message-helpers/index.js:79
    throw err
          ^
TypeError: <css input>:1:1: Cannot call method 'map' of undefined
    at <css input>:1:1
    at Root.Container.normalize (/Users/nullstyle/src/tmp/busted/node_modules/postcss/lib/container.js:451:25)
    at Root.normalize (/Users/nullstyle/src/tmp/busted/node_modules/postcss/lib/root.js:44:45)
    at Root.Container.insertBefore (/Users/nullstyle/src/tmp/busted/node_modules/postcss/lib/container.js:306:20)
    at insertRules (/Users/nullstyle/src/tmp/busted/node_modules/postcss-import/index.js:217:17)
    at readImportedContent (/Users/nullstyle/src/tmp/busted/node_modules/postcss-import/index.js:185:3)
    at readAtImport (/Users/nullstyle/src/tmp/busted/node_modules/postcss-import/index.js:150:3)
    at transformAtImport (/Users/nullstyle/src/tmp/busted/node_modules/postcss-import/index.js:86:7)
    at Object.tryCatch [as try] (/Users/nullstyle/src/tmp/busted/node_modules/postcss-import/node_modules/postcss-message-helpers/index.js:53:12)
    at checkAtRule (/Users/nullstyle/src/tmp/busted/node_modules/postcss-import/index.js:85:16)
    at /Users/nullstyle/src/tmp/busted/node_modules/postcss/lib/container.js:218:22
```

**Comment by nullstyle** [comment]
You can see an example project here: https://gist.github.com/nullstyle/c194cb3955d8226e18df

**Comment by MoOx** [comment]
I'm currently working on the upgrade.
postcss 4 seems to include a lot of changes on whitespace normalization which break a lot of tests.
Sometimes I get `foo{}` imported as `foo{}` (as expected), and sometimes I get the same foo.css file imported as `foo {}` (note the space). @ai any idea why I'm getting different result for the 2 first tests ?
Here I get normal import (on the first import) https://github.com/postcss/postcss-import/blob/master/test/fixtures/simple.css
Here I get the space https://github.com/postcss/postcss-import/blob/master/test/fixtures/no-duplicate.css.
I'll try to debug that this afternoon.

**Comment by ai** [comment]
How PostCSS stringify rules.

1. If rule has `between` property, PostCSS wwill use it.
2. If rules doesn’t have `between` property, PostCSS ill try to find other rule with `between` and use it.
3. If there is no rules in `between`, PostCSS will use default value.

When you `clone()` node, it will remove all whitespace properties: `before`, `after` and `between`.

When you use `moveTo()` it will remove `before` and `after`. Property `between` will be removed only if you move node to different root.

**Comment by MoOx** [comment]
@ai I'm not really using `moveTo()` for now, it's the same since I take a file that I import into another (so changing the root). Why is there a space added ?
Why moving a rule change the `between` ? It doesn't make sense (I admit that for after/before it does).

**Comment by ai** [comment]
Check does your node has own `between`. If you clone it, property should be deleted. So value is depend on how PostCSS find sample node.

**Comment by MoOx** [comment]
I just [read a file](https://github.com/postcss/postcss-import/blob/74d74fdd5a72a66eb5f53bbef95759e691843340/index.js#L180) then, [move the rules via insertBefore](https://github.com/postcss/postcss-import/blob/74d74fdd5a72a66eb5f53bbef95759e691843340/index.js#L217). So no clone.
It's just weird to see a `between` value to be created when none is given.

**Comment by MoOx** [comment]
I'm guessing in a way, you are [cloning all moved childs](https://github.com/postcss/postcss/blob/bb50a6eccf32fdfbc92b2106a1f8abf4df66050e/lib/container.js#L438) when you normalize. That's probably why I lose by empty `between`.
I'll need to make weird hack to make a import to be like it should :(

I'm really confused by this approach of updating the body of a rule that is just being "moved". I'm probably missing something, but I can't find a simple way to make a `foo{}` to be `foo{}` again in 2nd test. So weird.

**Comment by ai** [comment]
Idea behind PostCSS is to take a new code style, when you move rule to new root.

So, when you move `a{}` to `b { }`, in current PostCSS logic `between` should be cleaned and have value from new root `" "`.

**Comment by ai** [comment]
Second thing is that `normalize()` *must* to clone nodes. You can’t take node from root and insert it to other, because `root1.nodes` and `root2.nodes` will contains same node.

So, wwhen you move node from ont root to other to *must* to `clone()` them. Method `insertBefore()` have a check if you forgot to clone node. It will clone node for you.

When you clone node, this node will loose every whitespace property. Because it should take new code style from other nodes in new root.

**Comment by philipwalton** [comment]
@ai what you're saying makes sense, but I think the logic for determining the code style of the new root is a little weird. Consider a test case similar to what @MoOx mentions in https://github.com/postcss/postcss/issues/169

```css
@import "ignore.css" (min-width: 25em);

.foo {
  display: none;
}
```

What should the `between` property be in the above code? I'd argue it should be `' '` because that seems to be the style, but it's actually going to be `''` based on the current logic of PostCSS:

```js
postcss().process('@import "ignore.css" (min-width: 25em);').root.nodes[0].between;
// ouputs ''
```

From what I can tell with a quick step-through, it seem like [the logic](https://github.com/postcss/postcss/blob/7e1b183d841a426dfa0260d7e8188539a6a126d3/lib/node.js#L355-L358) in `Node#style` doesn't account for non-declaration rules like `@import` that don't have a "between" concept.

**Comment by ai** [comment]
@philipwalton yeap, at-rule without body doesn’t have `between` property. Only `afterName` property.

**Comment by ai** [comment]
@philipwalton and I totally agree that it is very strange, that `root.nodes[0].between == ''` O_o

If you properly clone node it should be `undefined`, because cloned node loose its code style to take code style from new root.

**Comment by MoOx** [comment]
Closed by 19604ea91cbfca15c09b9250a6ecbe7680f7745a
